### PR TITLE
Make slice() arguments optional

### PR DESCRIPTION
**Pull request by froomzy**
Changed slice to take Optional[int] rather than just straight int. Fix for #832.

### CODE REVIEW

File: stdlib/3/builtins.pyi [hunk 1]
```diff
@@ -472,7 +472,7 @@ class slice:
     start = 0
     step = 0
     stop = 0
-    def __init__(self, start: int, stop: int = 0, step: int = 0) -> None: ...
+    def __init__(self, start: Optional[int], stop: Optional[int] = 0, step: Optional[int] = 0) -> None: ...
```
Comment (ambv) [inline stdlib/3/builtins.pyi]: Please change the `step`'s default value to be `None`. In fact, it *cannot* be 0 :) Same for `stop`, the default value is `None`, behavior is different if 0 is given.

Comment (gvanrossum) [inline stdlib/3/builtins.pyi]: The default really should be `...`, in stubs the value of the default is irrelevant, only its presence is used.

Comment (ambv) [inline stdlib/3/builtins.pyi]: I can see how PyCharm could use actual defaults to help with auto-completion but yes, ... is indeed the canonical default value.

**Comment by gvanrossum** [comment]
I wonder if the `.start`, `.stop` and `.step` *attributes* should also be made `Optional`? OTOH that may cause breakage of user code. See discussions here:
- python/mypy#2410
- python/typing#159
But it would be a lie to leave them just `int` if the arguments are `Optional`.

**Comment by ambv** [comment]
From looking at this it seems to me adding `Slice[T, S, R]` to `typing` is the correct way to go. So in the mean time, should we accept this as an incremental (if incomplete) improvement for the common case?

**Comment by gvanrossum** [comment]
I expect some issues with existing code needing new markup when we make it
generic, so I think we can go ahead with this intermediate version first.

**Comment by ambv** [comment]
OK, given that the 0 default value is ignored anyway, I'm just going to merge it.

**Comment by ambv** [comment]
Thank you, @froomzy ✨ 🍰 ✨

**Pull request merged**